adapters/http4s/src/test/scala/caliban/Http4sAdapterSpec.scala

```diff
@@ -0,0 +1,210 @@
+      meta.map(_.path.toAbsolutePath.toString).getOrElse(""),
+      meta.map(_.fileName).getOrElse(""),
+      meta.flatMap(_.contentType).getOrElse("")
+    )
+
+  def uploadFiles(files: List[Upload]): ZIO[Uploads with Blocking, Throwable, List[TestAPI.File]] =
+    ZIO.collectAllPar(
+      for {
+        file <- files
+      } yield for {
+        bytes <- file.allBytes
+        meta  <- file.meta
+      } yield TestAPI.File(
+        Service.hex(Service.sha256(bytes.toArray)),
+        meta.map(_.path.toAbsolutePath.toString).getOrElse(""),
+        meta.map(_.fileName).getOrElse(""),
+        meta.flatMap(_.contentType).getOrElse("")
+      )
+    )
+
+  def sha256(b: Array[Byte]) =
+    MessageDigest.getInstance("SHA-256").digest(b)
+
+  def hex(b: Array[Byte]): String =
+    String.format("%032x", new BigInteger(1, b))
+}
+
+case class UploadFileArgs(file: Upload)
+case class UploadFilesArgs(files: List[Upload])
+
+object TestAPI extends GenericSchema[Blocking with Uploads with Console with Clock] {
+  val api: GraphQL[Blocking with Uploads with Console with Clock] =
+    graphQL(
+      RootResolver(
+        Queries(args => UIO("stub")),
+        Mutations(args => Service.uploadFile(args.file), args => Service.uploadFiles(args.files))
+      )
+    )
+
+  implicit val uploadFileArgsSchema = gen[UploadFileArgs]
+  implicit val mutationsSchema      = gen[Mutations]
+  implicit val queriesSchema        = gen[Queries]
+
+  case class File(hash: String, path: String, filename: String, mimetype: String)
+
+  case class Queries(stub: Unit => UIO[String])
+
+  case class Mutations(
+    uploadFile: UploadFileArgs => ZIO[Blocking with Uploads, Throwable, File],
+    uploadFiles: UploadFilesArgs => ZIO[Blocking with Uploads, Throwable, List[File]]
+  )
+}
+
+object Http4sAdapterSpec extends DefaultRunnableSpec {
+  type R = Console with Clock with Blocking with Random with Uploads
+  implicit val runtime: Runtime[R] =
+    Runtime.unsafeFromLayer(
+      Console.live ++ Clock.live ++ Blocking.live ++ Random.live ++ Uploads.empty,
+      Platform.default
+    )
+
+  val blocker = Blocker.liftExecutionContext(runtime.platform.executor.asEC)
+
+  val apiLayer: Layer[Throwable, Has[Server[RIO[R, *]]]] =
+    (for {
+      interpreter <- TestAPI.api.interpreter.toManaged_
+      server      <- BlazeServerBuilder(runtime.platform.executor.asEC)
+                       .bindHttp(8089, "127.0.0.1")
+                       .withHttpApp(
+                         (Http4sAdapter.makeHttpUploadService(
+                           interpreter,
+                           Paths.get(System.getProperty("java.io.tmpdir")),
+                           blocker
+                         ) <+> Http4sAdapter
+                           .makeHttpService(interpreter)).orNotFound
+                       )
+                       .resource
+                       .toManagedZIO
+    } yield server).toLayer
+      .asInstanceOf[Layer[Throwable, Has[Server[RIO[R, *]]]]]
```

Comment:
Ouf, that smells, but I’m not that familiar with this part of the ZIO Api, any directions here are appreciated 🚀

Reply:
Yeah, that's because `.resource` annoyingly returns a `Resource[F, Server[F]]` which doesn't play well with ZIO where we'd want two different F.

What you can do is change the signature of `apiLayer` to
```scala
val apiLayer: ZLayer[R, Throwable, Has[Server[RIO[R, *]]]]
```

Then make specLayer as:
```scala
val specLayer = ZLayer.requires[ZEnv] ++ Uploads.empty >>> apiLayer
```
which is providing `Uploads` and keep the rest as a requirement which will be provided by the test framework.

Reply:
@ghostdogpr ha, cool; works like a charm!